src/meepgeom.cpp

```diff
@@ -307,6 +308,121 @@ bool is_metal(meep::field_type ft, const material_type *material) {
+    signflip_dx = true;
+  }
+  if (dy != fabs(dy)) {
+    dy = fabs(dy);
+    signflip_dy = true;
+  }
+  if (dz != fabs(dz)) {
+    dz = fabs(dz);
+    signflip_dz = true;
+  }
+
+  /* define a macro to give us data(x,y,z) on the grid, 
+     in row-major order: */
+#define D(x, y, z) (data[(((x)*ny + (y)) * nz + (z)) * stride])
+
+  double du_dx = (signflip_dx ? -1.0 : 1.0) *
+    (((-D(x1, y1, z1) + D(x2, y1, z1)) * (1.0 - dy) +
+      (-D(x1, y2, z1) + D(x2, y2, z1)) * dy) * (1.0 - dz) +
+     ((-D(x1, y1, z2) + D(x2, y1, z2)) * (1.0 - dy) +
+      (-D(x1, y2, z2) + D(x2, y2, z2)) * dy) * dz);
+  double du_dy = (signflip_dy ? -1.0 : 1.0) *
+    ((-(D(x1, y1, z1) * (1.0 - dx) + D(x2, y1, z1) * dx) +
+      (D(x1, y2, z1) * (1.0 - dx) + D(x2, y2, z1) * dx)) * (1.0 - dz) +
+     (-(D(x1, y1, z2) * (1.0 - dx) + D(x2, y1, z2) * dx) +
+      (D(x1, y2, z2) * (1.0 - dx) + D(x2, y2, z2) * dx)) * dz);
+  double du_dz = (signflip_dz ? -1.0 : 1.0) *
+    (-((D(x1, y1, z1) * (1.0 - dx) + D(x2, y1, z1) * dx) * (1.0 - dy) +
+       (D(x1, y2, z1) * (1.0 - dx) + D(x2, y2, z1) * dx) * dy) +
+     ((D(x1, y1, z2) * (1.0 - dx) + D(x2, y1, z2) * dx) * (1.0 - dy) +
+      (D(x1, y2, z2) * (1.0 - dx) + D(x2, y2, z2) * dx) * dy));
+
+#undef D
+
+  gradient.set_direction(meep::X, du_dx);
+  gradient.set_direction(meep::Y, du_dy);
+  gradient.set_direction(meep::Z, du_dz);
+
+  return (abs(gradient) < 1e-8) ? zero_vec(dim) : gradient/abs(gradient);
+}
+
+meep::vec matgrid_grad(vector3 p, geom_box_tree tp, int oi, material_data *md) {
+  meep::vec gradient(zero_vec(dim));
+  int matgrid_val_count = 0;
+
+  if (md->material_grid_kinds == material_data::U_MIN ||
+      md->material_grid_kinds == material_data::U_PROD)
+    meep::abort("%s:%i:matgrid_grad does not support overlapping grids with U_MIN or U_PROD\n",__FILE__,__LINE__);
+
+  // iterate through object tree at current point
+  if (tp) {
+    do {
+      gradient += material_grid_grad(to_geom_box_coords(p, &tp->objects[oi]),
+                                     (material_data *)tp->objects[oi].o->material);
+      if (md->material_grid_kinds == material_data::U_DEFAULT) break;
+      ++matgrid_val_count;
+      tp = geom_tree_search_next(p, tp, &oi);
+    } while (tp && is_material_grid((material_data *)tp->objects[oi].o->material));
+  }
+  // perhaps there is no object tree and the default material is a material grid
+  if (!tp && is_material_grid(&default_material)) {
+    p.x = geometry_lattice.size.x == 0 ? 0
+                                       : 0.5 + (p.x - geometry_center.x) / geometry_lattice.size.x;
+    p.y = geometry_lattice.size.y == 0 ? 0
+                                       : 0.5 + (p.y - geometry_center.y) / geometry_lattice.size.y;
+    p.z = geometry_lattice.size.z == 0 ? 0
+                                       : 0.5 + (p.z - geometry_center.z) / geometry_lattice.size.z;
+    gradient = material_grid_grad(p, (material_data *)default_material);
+    ++matgrid_val_count;
+  }
```

Comment:
This code to compute `matgrid_val_count` seems duplicated; can we factor it out into a single function?

Reply:
The only duplicated code from `matgrid_grad` and `matgrid_val` that could be reasonably separated out was the coordinate mapping for the case of the default material.